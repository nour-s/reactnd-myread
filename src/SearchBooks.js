import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import BookCase from "./BookCase";

class SearchBooks extends React.Component {
	constructor(props) {
		super(props);
		this.state = { searchQuery: props.initQuery || "" };
	}

	componentWillUnmount() {
		this.props.onClearSearch();
	}

	componentDidMount() {
		//If the search component was called with a search query specified already
		if (this.props.initQuery) {
			//Let's run the search for the first time
			this.props.onSearch(this.props.initQuery);
		}
	}

	// A handler for the search text box
	onValueChange = e => {
		var query = e.target.value;
		this.setState({ searchQuery: query });
		this.props.onSearch(query);
	};

	render() {
		return (
			<div className="search-books">
				<div className="search-books-bar">
					<Link className="close-search" to={{ pathname: "/" }}>
						Close{" "}
					</Link>
					<div className="search-books-input-wrapper">
						{/*
                  NOTES: The search from BooksAPI is limited to a particular set of search terms.
                  You can find these search terms here:
                  https://github.com/udacity/reactnd-project-myreads-starter/blob/master/SEARCH_TERMS.md

                  However, remember that the BooksAPI.search method DOES search by title or author. So, don't worry if
                  you don't find a specific author or title. Every search is limited by search terms.
                */}
						<input
							type="text"
							onChange={this.onValueChange}
							placeholder="Search by title or author"
							value={this.state.searchQuery}
						/>
					</div>
				</div>
				<div className="search-books-results">
					{this.props.shelfs.length > 0 && (
						<BookCase
							shelfs={this.props.shelfs}
							onMoveBook={this.props.onMoveBook}
						/>
					)}
					<ol className="books-grid" />
				</div>
			</div>
		);
	}
}

SearchBooks.propTypes = {
	initQuery: PropTypes.string,
	shelfs: PropTypes.array,
	onMoveBook: PropTypes.func,
	onSearch: PropTypes.func,
	onClearSearch: PropTypes.func
};

export default SearchBooks;
