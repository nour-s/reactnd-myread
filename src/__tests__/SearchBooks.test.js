// Auto-generated do not edit

/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-undef */
import React from "react";
import renderer from "react-test-renderer";
import SearchBooks from "../SearchBooks";
import { MemoryRouter } from "react-router";
import { mount } from "enzyme";

describe("SearchBooks test", () => {
	let onMoveBook;
	let onSearch;
	let onClearSearch;
	let rawComponent;

	let shelfs = [
		{ id: "shelf 1", books: [{ id: 1, title: "Title 1" }] },
		{ id: "shelf 2", books: [{ id: 2, title: "Title 2" }] }
	];

	beforeEach(() => {
		onMoveBook = jest.fn();
		onSearch = jest.fn();
		onClearSearch = jest.fn();

		rawComponent = (
			<MemoryRouter>
				<SearchBooks
					initQuery={"anyQuery"}
					shelfs={shelfs}
					onMoveBook={onMoveBook}
					onSearch={onSearch}
					onClearSearch={onClearSearch}
				/>
			</MemoryRouter>
		);
	});

	it("SearchBooks should match snapshot", () => {
		const component = renderer.create(rawComponent);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("SearchBooks should have a link to main page", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);

		expect(component.find("Link").prop("to").pathname).toBe("/");
	});

	it("SearchBooks should have a BookCase component with proper props", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);

		expect(component.find("BookCase")).toBeDefined();
		expect(component.find("BookCase").prop("shelfs")).toBe(shelfs);
		expect(component.find("BookCase").prop("onMoveBook")).toBe(onMoveBook);
	});

	it("SearchBooks should have the text box populated with default query", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);
		var input = component.find("input");
		expect(input.prop("value")).toBe("anyQuery");
	});

	it("SearchBooks should have the text box populated with default query", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);
		var input = component.find("input");
		input.simulate("change", { target: { value: "newQuery" } });
		expect(onSearch).toBeCalledWith("newQuery");
	});

	it("SearchBooks should clear the search text box when unmounted", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);
		component.unmount();
		expect(onClearSearch).toBeCalled();
	});
});
