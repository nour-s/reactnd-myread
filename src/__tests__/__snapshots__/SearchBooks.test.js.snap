// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBooks test SearchBooks should match snapshot 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      onClick={[Function]}
    >
      Close
       
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value="anyQuery"
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <div
      className="list-books"
    >
      <div
        className="list-books-title"
      >
        <h1>
          MyReads
        </h1>
      </div>
      <div
        className="list-books-content"
      >
        <div>
          <div
            className="bookshelf"
          >
            <h2
              className="bookshelf-title"
            />
            <div
              className="bookshelf-books"
            >
              <ol
                className="books-grid"
              >
                <li>
                  <div
                    className="book"
                  >
                    <div
                      className="book-top"
                    >
                      <div
                        className="book-cover"
                        style={
                          Object {
                            "backgroundImage": "url()",
                            "height": 193,
                            "width": 128,
                          }
                        }
                      />
                      <div
                        className="book-shelf-changer"
                      >
                        <select
                          onChange={[Function]}
                          value="none"
                        >
                          <option
                            disabled={true}
                            value="move"
                          >
                            Move to...
                          </option>
                          <option
                            disabled={false}
                            value="currentlyReading"
                          >
                            Currently Reading
                          </option>
                          <option
                            disabled={false}
                            value="read"
                          >
                            Read
                          </option>
                          <option
                            disabled={false}
                            value="none"
                          >
                            None
                          </option>
                        </select>
                      </div>
                    </div>
                    <div
                      className="book-title"
                    >
                      Title 1
                    </div>
                  </div>
                </li>
              </ol>
            </div>
          </div>
          <div
            className="bookshelf"
          >
            <h2
              className="bookshelf-title"
            />
            <div
              className="bookshelf-books"
            >
              <ol
                className="books-grid"
              >
                <li>
                  <div
                    className="book"
                  >
                    <div
                      className="book-top"
                    >
                      <div
                        className="book-cover"
                        style={
                          Object {
                            "backgroundImage": "url()",
                            "height": 193,
                            "width": 128,
                          }
                        }
                      />
                      <div
                        className="book-shelf-changer"
                      >
                        <select
                          onChange={[Function]}
                          value="none"
                        >
                          <option
                            disabled={true}
                            value="move"
                          >
                            Move to...
                          </option>
                          <option
                            disabled={false}
                            value="currentlyReading"
                          >
                            Currently Reading
                          </option>
                          <option
                            disabled={false}
                            value="read"
                          >
                            Read
                          </option>
                          <option
                            disabled={false}
                            value="none"
                          >
                            None
                          </option>
                        </select>
                      </div>
                    </div>
                    <div
                      className="book-title"
                    >
                      Title 2
                    </div>
                  </div>
                </li>
              </ol>
            </div>
          </div>
        </div>
      </div>
      <div
        className="open-search"
      />
    </div>
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
